### PR TITLE
New Python unit tests for predicates and predicate libraries

### PR DESCRIPTION
**Pull request by benjaminulmer**
`test_hardware_library_merge_dups` motivated by #1280.

Other tests added to help ensure fallback to Device Efficiency kernels when a CU efficiency kernel cannot be found happens correctly (see #1279).

### CODE REVIEW

File: Tensile/Tests/unit/test_HardwarePredicates.py [hunk 1]
```diff
@@ -82,3 +83,21 @@ def test_predicate_library_merge(libraries):
...
+
+    return [copy.deepcopy(libs) for libs in itertools.permutations(objs)]
+
+@pytest.mark.parametrize("libraries", hardware_library_objects_dups())
+def test_hardware_library_merge_dups(libraries):
+    lib = libraries[0]
+    for lib2 in libraries[1:]:
+        lib.merge(lib2)
+
+    assert len(lib.rows) == 4
```
Comment (JeremyAdamHart) [inline Tensile/Tests/unit/test_HardwarePredicates.py:103]: This will probably be good enough, but it might be more reliable to test against the expected result, since it's possible for the merge  function to remove the wrong library and still pass.

File: Tensile/Tests/unit/test_PerfMetricPredicates.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+    return [copy.deepcopy(libs) for libs in itertools.permutations(objs)]
+
+@pytest.mark.parametrize("libraries", perf_metric_library_objects_dups())
+def test_perf_metric_library_merge_dups(libraries):
+    lib = libraries[0]
+    for lib2 in libraries[1:]:
+        lib.merge(lib2)
+
+    assert len(lib.rows) == 2
```
Comment (JeremyAdamHart) [inline Tensile/Tests/unit/test_PerfMetricPredicates.py:67]: Same comment here

**Review by JeremyAdamHart — COMMENTED** [review]
Aside from the two comments, lgtm!

File: Tensile/Tests/unit/test_HardwarePredicates.py (hunk 1 above)
Comment (benjaminulmer) [reply Tensile/Tests/unit/test_HardwarePredicates.py:103]: Sure. It's a pretty unlikely edge case, but testing to ensure against it should be simple enough.

**Review by cgmillette — CHANGES_REQUESTED** [review]
While it's good that we're testing the general expectations in terms of resulting library row count etc., the verification should include the specific expected outputs as well. Like Jeremy mentioned, we should check the specific expected output because we have an explicit input case. Gotta test it like a 'black box' and not trust it will do anything correctly :)

**Comment by benjaminulmer** [comment]
> While it's good that we're testing the general expectations in terms of resulting library row count etc., the verification should include the specific expected outputs as well. Like Jeremy mentioned, we should check the specific expected output because we have an explicit input case. Gotta test it like a 'black box' and not trust it will do anything correctly :)

Agreed. I just updated this in the latest commit.

**Review by cgmillette — APPROVED** [review]
LGTM

**Review by JeremyAdamHart — APPROVED** [review]
Looks all good to me too!
